### PR TITLE
Storage interface

### PR DESCRIPTION
**Pull request by AlexPraefectus**
Storage is splitted into dynamic namespaces to allow logical data separation.
Each namespace consists of segments.
This decision will allow to:
* improve data persistence performance (track changed segments and sync only them to the disk)
* improve API operations performance because we can read from one segment and write to another in parallel
* avoid "stop-the-world" locks for namespace because each segment can be processed separately (by any task)

At this point storage is only in-memory and will be empty on process start.
It's done this way because it will allow code which uses storage work with
stable real api (with no fakes). And after this any persistence/housekeeping
can be added without worrying about developers who cannot do their part of work.

Minors:
* Java generics do not support primitives (reason to add Container class)
* renamed bigtable to storage because proposed implementation doesn't care about how bigtable works

### CODE REVIEW

File: src/main/java/edu/kpi/testcourse/storage/StorageImpl.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import edu.kpi.testcourse.storage.namespace.Namespace;
+import edu.kpi.testcourse.storage.namespace.NamespaceImpl;
+import java.util.ArrayList;
+
+/**
+ * Key-value storage implementation.
+ */
+public class StorageImpl implements Storage {
+
+  private final ArrayList<NamespaceImpl> namespaces;
```
Comment (OlehKyba) [inline src/main/java/edu/kpi/testcourse/storage/StorageImpl.java:12]: Why do you use an implementation class? Maybe we should use the interface here?

File: src/main/java/edu/kpi/testcourse/storage/namespace/HashParts.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  private static final int LOW_MASK = 0x0000ffff;
+
+  /**
+   * Splits hash of string to most and least significant bits.
+   *
+   * @param key string, self-descriptive.
+   *
+   * @return HashParts where low is 16 least significant bits and high is 16 most significant bits
+   */
+  static HashParts from(String key) {
```
Comment (OlehKyba) [inline src/main/java/edu/kpi/testcourse/storage/namespace/HashParts.java:34]: Maybe we should overload the constructor? In my opinion, it is more Java way.

File: src/main/java/edu/kpi/testcourse/storage/StorageImpl.java (hunk 1 above)
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/storage/StorageImpl.java:12]: Yup, You're completely right, I'll change this.

File: src/main/java/edu/kpi/testcourse/storage/namespace/HashParts.java (hunk 2 above)
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/storage/namespace/HashParts.java:34]: `HashParts` is some kind of data class with a bit of logic.
I think that It's better to only assign fields in the constructor and move the logic of creating one type (`HashParts`) from another (`String`) to another place. But it's not enough logic to introduce a separate builder class so I've moved it into the static method of `HashParts`

Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/storage/namespace/HashParts.java:34]: But if in Your opinion it's better to make the constructor look like `HashParts(String s)` I'll change this.

**Review by OlehKyba — APPROVED** [review]
:+1:

**Pull request merged**